### PR TITLE
ci(deployment): simplify deployment workflow

### LINKED ISSUE #1426: Deployment simplification
Context: [Our current deployment process](https://docs.google.com/document/d/15XFxCZ4Si6NW9abTi2PG7kM50WUshsbYW43dGRR81_8/edit#heading=h.ad60101oqy8o) is rather tedious, we want to reduce:
- the number of sources of truth for deployments
- the number of steps involved in a release

DoD:
- [ ] Research on existing deployment processes 
- [ ] Analyse the advantage/disadvantages and decide on an approach
- [ ] Write an RFC doc on the proposed process

### PR DESCRIPTION
**Pull request by stanleynguyen**
## Problem

We're current having a very tedious deployment process of:
- updating each sub-package's version
- creating a release branch
- create a release tag
- PR and merge the release branch into master (to trigger deploy to prod)
- PR and merge back release branch (with the release commit) to `develop`

Closes #1426 

## Solution

**TLDR;** New prod deployment process will be a one-step process of [creating a new release](https://github.com/opengovsg/postmangovsg/releases/new) on Github

**More deets**:
- Consolidating the git history for all deployment artefacts to use a single branch `master`.
- `master` will be set to the main branch and `develop` will be removed. All PRs will now target to `master` (Prod deployments won't be triggered on push to `master` anymore)
- To create deployment artefact and trigger deploy to prod, we will use semver tags (e.g. `v1.1.1`) that can be done through [github UI](https://github.com/opengovsg/postmangovsg/releases/new) easily with release notes 
- Ideally, we want `staging` env to be built from a `master`-commit also and triggered using release candidate tags (e.g. `v1.1.1-rc01`), but AWS Amplify can't deploy based off tags (only branch for now) and we are also using our `staging` like a `dev` environment to test changes in real infras => We will keep staging deployment to status quo `push to "staging"` 
- Wrt the risk of AWS Amplify deploying the incorrect commit caused by master being ahead of release tag, it's mitigated by the fact that we're still doing scheduled release with the latest commit from master being released
- Wrt concerns about being able to view version in deployed environment from code, build systems nowadays are pretty accurate about the artefacts so we can rely on the commit hash that we add to the artefact names.

**Side Note**: I'm also taking this chance to bump our node version in CD to 16 and get rid of the problematic npm cache

## Deployment Checklist
- [ ] Do a final release with the old way to cap off and align the `master` and `develop` history
- [ ] Immediately create a new release with this updated flow after the previous step:
  - [ ] Switch main branch to `master` instead of `develop` (with all protek rules)
  - [ ] Create a semver tag from master to trigger deployment
- [ ] Remove `develop` branch after the next release (just to be sure)

### CODE REVIEW

**Comment by stanleynguyen** [comment]
Pending update to README

**Comment by prestonlimlianjie** [comment]
> `master` will be set to the main branch and develop will be removed. All PRs will now target to master (Prod deployments won't be triggered on push to master anymore)

How do we deploy our master branch code to the prod AWS environment?

**Comment by stanleynguyen** [comment]
As per our discussion, we will make a more drastic migration of our deployment workflows to `release small and release more frequently`. Will modify this PR to:
- Immediately release upon merging to master
- Update deployment guideline docs to contain information about:
  - implication of merging to main branch `master` and checklist before that
  - how to rollback on platforms EB, ECS, Beanstalks
- Manual Github Action workflow to rollback to a specific commit hash
- Update Github protection rule to require PR to be up-to-date with main branch `master`

**Comment by tan-zx** [comment]
Some questions for my own learning:
1. I understand that AWS Amplify is like Amazon's version of Firebase and we're using to deploy the frontend → it's my first time encountering Amplify in OGP products and is there any reason why Postman requires using Amplify? Based on your write-up, it seems like Amplify is a blocker for the deployment workflow you want to build? (I also don't fully what you mean by "the risk of AWS Amplify deploying the incorrect commit caused by master being ahead of release tag")


Separately, is it right to say that the new deployment process is:

**Comment by stanleynguyen** [comment]
Extra context: After the discussion, we're not going with a tag-based/semver-versioning deployment approach anymore but just deploy straight from `master` and version track using commit hash

> Is there any reason why Postman requires using Amplify?
It's for frontend hosting (not sure which ones u're familiar with but it's the same as Netlify, Surge.sh, Vercel)

> Based on your write-up, it seems like Amplify is a blocker for the deployment workflow you want to build → is this something we can change out
Not anymore as we decided to get rid of the tags as well

> I also don't fully what you mean by "the risk of AWS Amplify deploying the incorrect commit caused by master being ahead of release tag"
Let's say your commit history (from oldest to latest) is A->B->C. Because Amplify deploy based on branch (always latest commit), even if you tag B to be the intended version to be deployed, Amplify will be triggered to deploy on C instead

> Is this correct: deploy-frontend.yml runs ./scripts/amplify-deploy.sh, which in term makse use of the amplify.yml file as part of the aws apmlify command?
No, the script only trigger and wait for the Amplify build and deploy. Amplify build environment will make use of the file

2. Nothing too significant for us but u can check it out on their changelog
3. `db-migration.yml` is to manually migrate the db as well, but using the Github Action UI so:
  - We have a records of who ran what at what time
  - we won't have credentials lying around on unprotected environments
4. no we're not using tags based release. Any merge into master will be deployed immediately. Please refer to the workflows yml
